src/counter.rs

```diff
@@ -0,0 +1,63 @@
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+use std::sync::atomic::{AtomicIsize, Ordering};
+
+pub(crate) struct ConcurrentCounter {
+    base: AtomicIsize,
+    cells: Vec<AtomicIsize>,
+}
+
+impl ConcurrentCounter {
+    pub(crate) fn new() -> Self {
+        Self {
+            base: AtomicIsize::new(0),
+            cells: (0..crate::map::num_cpus())
+                .into_iter()
+                .map(|_| AtomicIsize::new(0))
+                .collect(),
+        }
+    }
+
+    pub(crate) fn add(&self, value: isize) {
+        let base = self.base.load(Ordering::SeqCst);
+        if self
+            .base
+            .compare_exchange(base, base + value, Ordering::SeqCst, Ordering::Relaxed)
+            .is_err()
+        {
+            // we experienced contention on base
+            loop {
+                let c = &self
+                    .cells
+                    .choose(&mut rand::rngs::SmallRng::from_entropy())
+                    .unwrap();
```

Comment:
This is maybe a super weird proposal, but what if we chose a cell based on `base % self.cells.len()` (maybe expressed better since we know `self.cells` is base-2-length? That way we don't need to sample randomness at all.

Reply:
I think it's actually a nicer way of selecting a cell without `rand`, with the added bonus that we can remove that dependency.